File: nodejs/lib/sms.js
const _ = require('lodash');
const soap = require('soap');
const { parseString } = require('xml2js');

class SMS {
  constructor({ endpoint, clientNo, clientPass } = {}) {
    this.endpoint = 'http://center.fibosms.com/service.asmx';
    this.clientNo = clientNo;
    this.clientPass = clientPass;
    this.serviceType = 1;
    this.client = null;
  }

  async GetClientBalance() {
    const response = await this.execute('GetClientBalance', {
      clientNo: this.clientNo,
      clientPass: this.clientPass,
      serviceType: 0
    });

    console.log('GetClientBalance > ', response);
    return response;
  }

  /**
   * 
   * @param {*} msg {{senderName, phoneNumber, smsMessage, smsGUID}}
   * @param {*} sendBy 
   */
  async sendSMS(msg, sendBy = 'soap') {
    const baseParam = {
      clientNo: this.clientNo,
      clientPass: this.clientPass,
      serviceType: 0,
    };
    if (_.isArray(msg)) {
      return Promise.all(msg.map(m => this.execute('SendMaskedSMS', {
        ...baseParam,
        ...m,
      }, sendBy)));
    }
    return this.execute('SendMaskedSMS', {
      ...baseParam,
      ...msg,
    }, sendBy);
  }

  async execute(service, params) {
    this.client = await this.createClient();
    if (!this.client) {
      return null;
    }
    const response = await this.client[`${service}Async`](params);
    return new Promise((resolve, reject) => {
      console.log('execute: ', params, ', response: ', response);
      const body = response[`${service}Result`];
      if (/^[\d]+$/.test(body)) {
        return resolve(body);
      }
      parseString(body, (error, data) => {
        if (error) return reject(error);
        resolve(data);
      })
    })
  }

  async createClient() {
    if (!this.client) {
      try {
        return soap.createClientAsync(`${this.endpoint}?wsdl`);
      } catch (error) {
        console.log('[Error] GetClientBalance > ', error);
      }
    }
  }

  toUri(service, query) {
    return `${this.endpoint}/${service}?${query}`;
  }

  buildQuery(params) {
    let query = {};
    if (params) {
      query = {
        ...query,
        ...params,
      }
    }
    return Object.keys(query).map(k => `${k}=${query[k]}`).join('&');
  }
}

module.exports = SMS;

File: java/lib/vn/fibo/sms/SMS.java
package vn.fibo.sms;

import java.util.Map;

import javax.xml.soap.*;

public class SMS {
  private String endpoint;
  private String clientNo;
  private String clientPass;
  private int serviceType;
  private Object client;

  public SMS(String endpoint, String clientNo, String clientPass, int serviceType) {
    this.endpoint = endpoint;
    this.clientNo = clientNo;
    this.clientPass = clientPass;
    this.serviceType = serviceType;
    this.client = null;
  }

  public SMS(String clientNo, String clientPass) {
    this.endpoint = "http://center.fibosms.com/service.asmx";
    this.clientNo = clientNo;
    this.clientPass = clientPass;
    this.serviceType = 1;
    this.client = null;
  }

  public boolean sendSMS(String message) {
    SOAPMessage response = this.callSoapWebService(this.toUri(), "SendMaskedSMS");
    // get the body
    SOAPBody soapBody = response.getSOAPBody();
    // find your node based on tag name
    NodeList nodes = soapBody.getElementsByTagName("Message");
    String result = null;
    Node node = nodes.item(0);
    result = node != null ? node.getTextContent() : "";
    return result == "Sending...";
  }

  private void createSoapEnvelope(SOAPMessage soapMessage) throws SOAPException {
    SOAPPart soapPart = soapMessage.getSOAPPart();

    String myNamespace = this.endpoint + "?wsdl";
    String myNamespaceURI = "http://www.webserviceX.NET/";

    // SOAP Envelope
    SOAPEnvelope envelope = soapPart.getEnvelope();
    envelope.addNamespaceDeclaration(myNamespace, myNamespaceURI);
  }

  private SOAPMessage callSoapWebService(String soapEndpointUrl, String soapAction) {
    SOAPConnection soapConnection;
    try {
      // Create SOAP Connection
      SOAPConnectionFactory soapConnectionFactory = SOAPConnectionFactory.newInstance();
      soapConnection = soapConnectionFactory.createConnection();

      // Send SOAP Message to SOAP Server
      SOAPMessage soapResponse = soapConnection.call(createSOAPRequest(soapAction), soapEndpointUrl);

      return soapResponse;
    } catch (Exception e) {
      System.err.println(
          "\nError occurred while sending SOAP Request to Server!\nMake sure you have the correct endpoint URL and SOAPAction!\n");
      e.printStackTrace();
    } finally {
      if (soapConnection) {
        soapConnection.close();
      }
    }
  }

  private SOAPMessage createSOAPRequest(String soapAction) throws Exception {
    MessageFactory messageFactory = MessageFactory.newInstance();
    SOAPMessage soapMessage = messageFactory.createMessage();

    createSoapEnvelope(soapMessage);

    MimeHeaders headers = soapMessage.getMimeHeaders();
    headers.addHeader("SOAPAction", soapAction);

    soapMessage.saveChanges();

    /* Print the request message, just for debugging purposes */
    System.out.println("Request SOAP Message:");
    soapMessage.writeTo(System.out);
    System.out.println("\n");

    return soapMessage;
  }

  private String toUri(String service, String query) {
    return String.format("%s/%s?%s", this.endpoint, service, query);
  }

  private String buildQuery(Map<String, String> params) {
    return params.entrySet().stream().map(p -> p.getKey() + "=" + p.getValue()).reduce((p1, p2) -> p1 + "&" + p2)
        .map(s -> "?" + s).orElse("");
  }
}
